chain/testing.go

```diff
@@ -367,14 +367,21 @@ func (f *Builder) GetTipSet(key types.TipSetKey) (types.TipSet, error) {
 	return types.NewTipSet(blocks...)
 }
 
-// FetchTipSets returns `recur` tipsets from `key` by following parent keys.
-func (f *Builder) FetchTipSets(ctx context.Context, key types.TipSetKey, recur int) ([]types.TipSet, error) {
+// FetchTipSets fetchs the tipset at `tsKey` from the fetchers blockStore backed by the Builder.
+func (f *Builder) FetchTipSets(ctx context.Context, key types.TipSetKey, done func(t types.TipSet) (bool, error)) ([]types.TipSet, error) {
 	var tips []types.TipSet
-	for i := 0; i < recur; i++ {
+	for {
 		tip, err := f.GetTipSet(key)
 		if err != nil {
 			return nil, err
 		}
+		ok, err := done(tip)
+		if err != nil {
+			return nil, err
+		}
+		if ok {
+			break
+		}
```

Comment:
Blocking: the last fetched tip should still be returned. I think the loop simplifies to `for !done {...}`.

Reply:
I have addressed the part about including the last fetched tipset in the result, however `done` will still need to be called inside the loop (as opposed to the `for {` declaration) since it operators on a `TipSet` and we don't yet have one at that point.